src/main/java/io/vertx/mqtt/impl/MqttConnection.java

```diff
@@ -0,0 +1,315 @@
+   */
+  public MqttConnection(VertxInternal vertx, Channel channel, ContextImpl context, Handler<MqttEndpoint> endpointHandler, TCPMetrics metrics) {
+    super(vertx, channel, context);
+
+    this.endpointHandler = endpointHandler;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public Object metric() {
+    return null;
+  }
+
+  @Override
+  protected void handleInterestedOpsChanged() {
+  }
+
+  /**
+   * Handle the MQTT message received by the remote MQTT client
+   *
+   * @param msg message to handle
+   */
+  synchronized void handleMessage(Object msg) {
+
+    // handling directly native Netty MQTT messages because we don't need to
+    // expose them at higher level (so no need for polyglotization)
+    if (msg instanceof io.netty.handler.codec.mqtt.MqttMessage) {
+
+      io.netty.handler.codec.mqtt.MqttMessage mqttMessage = (io.netty.handler.codec.mqtt.MqttMessage) msg;
+
+      switch (mqttMessage.fixedHeader().messageType()) {
+
+        case CONNECT:
+          handleConnect((MqttConnectMessage) msg);
+          break;
+
+        case PUBACK:
+
+          io.netty.handler.codec.mqtt.MqttPubAckMessage mqttPubackMessage = (io.netty.handler.codec.mqtt.MqttPubAckMessage) mqttMessage;
+          this.handlePuback(mqttPubackMessage.variableHeader().messageId());
+          break;
+
+        case PUBREC:
+
+          int pubrecMessageId = ((io.netty.handler.codec.mqtt.MqttMessageIdVariableHeader) mqttMessage.variableHeader()).messageId();
+          this.handlePubrec(pubrecMessageId);
+          break;
+
+        case PUBREL:
+
+          int pubrelMessageId = ((io.netty.handler.codec.mqtt.MqttMessageIdVariableHeader) mqttMessage.variableHeader()).messageId();
+          this.handlePubrel(pubrelMessageId);
+          break;
+
+        case PUBCOMP:
+
+          int pubcompMessageId = ((io.netty.handler.codec.mqtt.MqttMessageIdVariableHeader) mqttMessage.variableHeader()).messageId();
+          this.handlePubcomp(pubcompMessageId);
+          break;
+
+        case PINGREQ:
+
+          this.handlePingreq();
+          break;
+
+        case DISCONNECT:
+
+          this.handleDisconnect();
+          break;
+
+        default:
+
+          this.channel.pipeline().fireExceptionCaught(new Exception("Wrong message type " + msg.getClass().getName()));
+          break;
+
+      }
+
+      // handling mapped Vert.x MQTT messages (from Netty ones) because they'll be provided
+      // to the higher layer (so need for ployglotization)
+    } else {
+
+      if (msg instanceof MqttSubscribeMessage) {
+
+        this.handleSubscribe((MqttSubscribeMessage) msg);
+
+      } else if (msg instanceof MqttUnsubscribeMessage) {
+
+        this.handleUnsubscribe((MqttUnsubscribeMessage) msg);
+
+      } else if (msg instanceof MqttPublishMessage) {
+
+        this.handlePublish((MqttPublishMessage) msg);
+
+      } else {
+
+        this.channel.pipeline().fireExceptionCaught(new Exception("Wrong message type"));
```

Comment:
@vietj agree ! even if it's quite difficult or impossible using an Mqtt client from Eclipse Paho. I should write a simple test client which open TCP socket and then sends a first MQTT packet with an unknown code for the packet identifier (the first packet should be CONNECT, I should send something not existing). I'll try to do that.